### PR TITLE
Frontend, Auth, API tokens, CORS, Categories

### PR DESCRIPTION
**Pull request by BradyHodge**
## UI Implemented With EJS
- Built the frontend using ejs (Basicaly HTML with variables)
- Users can create, edit, and delete their tasks
- Users can sort and filter tasks
- Users can create and manage api tokens
- UI will refresh if fetch returns 401 (unauthorized) 

## Added Authentication to routes
- requireAuth redirects browser to /auth if not logged in
- requireAuthAPI returns 401 in json if not logged in
- Auth controller creates new user in database if the user has never logged in before
- Auth controller update database on user log in (in case username or whatever has changed)
- Task route only returns task associated with the user's id
- Require task to be associated with a user

## API tokens
- Set up the the require auth routes to check for ether an api token in the header or a sesson token
- Set up bearer tokens (for the api tokens)
- Tokens are stored in the database, associated with a spesic user
- Each token has a name, created date, and last used date

## CORS
- Validates the domain
- Requres https when in prod 

## Category
- Updated category to work with Oauth
- Updated categories to only return the ones associated with the user
- Updated to use the RequestHandler type
- Updated schema to better handle a task refrencing a category

### CODE REVIEW

File: src/middleware/authMiddleware.ts [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    req: Request,
+    res: Response,
+    next: NextFunction
+): Promise<void> => {
+    const userReq = req as AuthRequest;
+
+    // First check for API token in header
+    const authHeader = req.headers.authorization;
+    // Make sure the header exists and starts with 'Bearer ' (the api token type)
+    // idk if this is really necessary, but the dude on stackoverflow said to do it
```
Comment (carteraymond) [inline src/middleware/authMiddleware.ts:30]: lol

**Review by carteraymond — APPROVED** [review]
You have really shown alot of skill and dedication to your work. I am very impressed with your comments throughout as well as the readability. Even though I am not well versed in 0auth, it felt like I understood the flow in the logic well. Thank you so much for sticking it out with our group.

File: src/routes/task.ts [hunk 2]
```diff
@@ -1,104 +1,21 @@
 import express from 'express';
 import { create, getAll, getId, remove, update } from '../controllers/taskController';
+import { requireAuthAPI } from '../middleware/authMiddleware';
 
-const taskRouter = express.Router();
+const TaskRouter = express.Router();
```
Comment (carteraymond) [inline src/routes/task.ts]: You Capitalized My TaskRouter?

Comment (BradyHodge) [reply src/routes/task.ts]: Ummmmmmm..... idk I think I was still in C++ mode?

**Pull request merged**